### PR TITLE
Add apiVersion to specific message filter onXyz[Request|Response] methods

### PR DESCRIPTION
**Pull request by robobario**
### Type of change

- Enhancement / new feature

### Description

Adds apiVersion to specific message filter onXyz[Request|Response] methods.

Why:
Currently we do not have a way to access the apiVersion for Responses. The apiVersion is in the request headers but so far we have not made it available to response filters. This is a problem when we want to construct response data in the proxy, we need to know which fields can be serialized at the apiVersion used by the client. Also we have started adding code which is trying to sniff the version by checking for non-null fields, sometimes it might be cleaner to switch on the apiVersion instead.

I opted to add it to the signatures for requests and responses for symmetry despite apiVersion also being present in the request header data.

edit: alternatively it could be contextual data delivered via KrpcContext?

### CODE REVIEW

File: api/kroxylicious-filter-api/src/main/templates/Kproxy/Filter.ftl [hunk 1]
```diff
@@ -63,6 +63,6 @@ public interface ${filterClass} extends KrpcFilter {
      * @param ${msgType} The KRPC message to handle.
      * @param context The context.
      */
-    void on${messageSpec.name}(<#if messageSpec.type?lower_case == 'response'>ResponseHeaderData<#else>RequestHeaderData</#if> header, ${dataClass} ${msgType}, KrpcFilterContext context);
+    void on${messageSpec.name}(short apiVersion, <#if messageSpec.type?lower_case == 'response'>ResponseHeaderData<#else>RequestHeaderData</#if> header, ${dataClass} ${msgType}, KrpcFilterContext context);
```
Comment (robobario) [inline api/kroxylicious-filter-api/src/main/templates/Kproxy/Filter.ftl:66]: need to add it to javadoc

**Comment by SamBarker** [comment]
My initial reaction was we should include it on both the request and response methods in the API. 

However this comment:
>edit: alternatively it could be contextual data delivered via KrpcContext?

Triggered another thought: Should the `onResponse` methods have access to the request or the request Headers in general?

It moves us away from providing "pure" transformation filters into offering Request Response aware filters. I'm not really clear what the implications of that are or what use-cases that helps us support.

To answer the actual question though, No I don't think it should be part of the context as relates to a specific request response pair not general interactions with the proxy itself.

**Comment by k-wall** [comment]
> 
> Triggered another thought: Should the `onResponse` methods have access to the request or the request Headers in general?

Filter authors can achieve this with the existing API. We already provide the guarantee that a filter instance is not re-used across channels. The filter author could implement a filter implements `XRequestFilter` and `XResponseFilter`.  On the request side, they cache whatever details they need from the request as filter instance state.   The response side then uses the information.   Admittedly, that is more work for the filter author.

If `onResponse` API were to offer access to the *full request*, that would mean that any XResponseFilter implementation would always impose the cost the decoding the request.  That would disadvantage use-cases that don't need the request (performance hit). 

If `onResponse` API were to offer access to the *request header only*, there will would be a `KafkaRequestDecoder#readHeader` performance hit imposed on those use-cases.   Is it worth it?  The only thing it would give the response filter author beyond the `apiVersion` would be the `clientId` and the request's `_unknownTaggedFields`.









> 
> It moves us away from providing "pure" transformation filters into offering Request Response aware filters. I'm not really clear what the implications of that are or what use-cases that helps us support.
> 
> To answer the actual question though, No I don't think it should be part of the context as relates to a specific request response pair not general interactions with the proxy itself.

**Review by k-wall — APPROVED** [review]
+1 to adding the apiVersion to the API.  thanks @robobario

**Comment by SamBarker** [comment]
> Filter authors can achieve this with the existing API. We already provide the guarantee that a filter instance is not re-used across channels. The filter author could implement a filter implements XRequestFilter and XResponseFilter. On the request side, they cache whatever details they need from the request as filter instance state. The response side then uses the information. Admittedly, that is more work for the filter author.

Yeah it's more work and potentially often repeated. 

Note I'm thinking out-loud rather than advocating for it at this point. I'm trying to envisage the most helpful api for filter authors which makes the most common tasks simple.

Admittedly the sample size is small and potentially biased but most of the filters we have talked about so far are symmetric in that they apply to both the request and response. So I'm thinking about optimising for those cases.

> If onResponse API were to offer access to the full request, that would mean that any XResponseFilter implementation would always impose the cost the decoding the request. That would disadvantage use-cases that don't need the request (performance hit).

That's a good point, but if we passed it in with a wrapper we can lazily decode as required (or return the pre-decoded value from the request) I think we have the state in the correlation manager anyway. Having said that what version of the request to provide to the response? The version the client sent or the version forwarded on? 

> The only thing it would give the response filter author beyond the apiVersion would be the clientId and the request's _unknownTaggedFields.

I think there is a fair chance that those would be interesting to a subset of filter authors. IIRC the schema registry uses that to mark schema coordinates and that's potentially useful in the response. But maybe we are talking about rare enough cases that caching the interesting data in onRequest is good enough.

**Comment by robobario** [comment]
> That's a good point, but if we passed it in with a wrapper we can lazily decode as required (or return the pre-decoded value from the request) I think we have the state in the correlation manager anyway. Having said that what version of the request to provide to the response? The version the client sent or the version forwarded on?

Currently we are constrained in the Filters to only forward requests and responses at the `apiVersion` requested by the client (after it discovers the supported apiVersions from the broker+kroxy). So `apiVersion` in the `onXyzRequest|Response` signatures means the `apiVersion` from the client. If you tried to set fields incompatible with that version on the data object it will fail at encode time.

An exception is `sendRequest`, we should make the apiVersion configurable here since the request is initiated by the proxy.

I wonder if these are acceptable limitations. It's conceivable that we could, at the proxy, want to do things like forward a request to the broker at a higher apiVersion than the client supports and then downgrade again at the proxy.

**Review by SamBarker — APPROVED** [review]
>  > The only thing it would give the response filter author beyond the apiVersion would be the clientId and the request's _unknownTaggedFields.

> I think there is a fair chance that those would be interesting to a subset of filter authors. IIRC the schema registry uses that to mark schema coordinates and that's potentially useful in the response. But maybe we are talking about rare enough cases that caching the interesting data in onRequest is good enough.

No, I'm getting Message headers and request headers confused. `short apiVersion` is good enough for me.

**Pull request merged**